dbt/compilation.py

```diff
@@ -261,6 +261,22 @@ def get_all_projects(self):
 
         return all_projects
 
+    def _check_resource_uniqueness(cls, flat_graph):
+        nodes = flat_graph['nodes']
+        names_resources = {}
+
+        for resource, attribs in nodes.items():
+            if attribs.get('resource_type') not in NodeType.refable():
+                continue
+
+            name = attribs['name']
+            existing_name = names_resources.get(name)
+            if existing_name is not None:
+                dbt.exceptions.raise_duplicate_resource_name(resource,
+                                                             existing_name)
+
+            names_resources[name] = resource
```

Comment:
do we have logic like this anywhere else?

Reply:
This code previously existed in the loader, but I ripped it out and put it in the compiler class. I liked the idea of the Loader just returning a flat graph and not validating uniqueness or other business logic constraints. It feels like that will be important if we expose the loader through our API

Reply:
We do something similar here: https://github.com/fishtown-analytics/dbt/blob/development/dbt/parser.py#L290